core/grant/service.go

```diff
@@ -480,6 +484,132 @@ func (s *Service) ImportFromProvider(ctx context.Context, criteria ImportFromPro
+	if err := criteria.Validate(); err != nil {
+		return fmt.Errorf("invalid criteria: %w", err)
+	}
+	startDate := time.Now().Add(-criteria.Period)
+
+	provider, err := s.providerService.GetByID(ctx, criteria.ProviderID)
+	if err != nil {
+		return fmt.Errorf("getting provider details: %w", err)
+	}
+
+	s.logger.Info(fmt.Sprintf("getting active grants for provider %q", provider.URN))
+	grants, err := s.List(ctx, domain.ListGrantsFilter{
+		Statuses:      []string{string(domain.GrantStatusActive)}, // TODO: evaluate later to use status_in_provider
+		ProviderTypes: []string{provider.Type},
+		ProviderURNs:  []string{provider.URN},
+		CreatedAtLte:  startDate,
+	})
+	if err != nil {
+		return fmt.Errorf("listing active grants: %w", err)
+	}
+	if len(grants) == 0 {
+		s.logger.Info(fmt.Sprintf("no active grants found for provider %q", provider.URN))
+		return nil
+	}
+	grantIDs := getGrantIDs(grants)
+	s.logger.Info(fmt.Sprintf("found %d active grants for provider %q", len(grants), provider.URN), "grant_ids", grantIDs)
+
+	var uniqueAccountIDs []string
+	uniqueAccountIDsMap := map[string]bool{}
+	for _, g := range grants {
+		if _, ok := uniqueAccountIDsMap[g.AccountID]; !ok {
+			uniqueAccountIDs = append(uniqueAccountIDs, g.AccountID)
+			uniqueAccountIDsMap[g.AccountID] = true
+		}
+	}
+
+	s.logger.Info(fmt.Sprintf("getting activities for provider %q", provider.URN))
+	activities, err := s.providerService.ListActivities(ctx, *provider, domain.ListActivitiesFilter{
+		AccountIDs:   uniqueAccountIDs,
+		TimestampGte: &startDate,
+	})
+	if err != nil {
+		return fmt.Errorf("listing activities: %w", err)
+	}
+
+	grantsPointer := make([]*domain.Grant, len(grants))
+	for i, g := range grants {
+		g := g
+		grantsPointer[i] = &g
+	}
+	if err := s.providerService.CorrelateGrantActivities(ctx, *provider, grantsPointer, activities); err != nil {
+		return fmt.Errorf("correlating grant activities: %w", err)
+	}
+
+	s.logger.Info("checking grants dormancy...")
+	var dormantGrants []*domain.Grant
+	var dormantGrantsIDs []string
+	var dormantGrantsByOwner = map[string][]*domain.Grant{}
+	for _, g := range grantsPointer {
+		if len(g.Activities) == 0 {
+			g.ExpirationDateReason = fmt.Sprintf("%s: %s", domain.GrantExpirationReasonDormant, criteria.RetainDuration)
+			newExpDate := time.Now().Add(criteria.RetainDuration)
+			g.ExpirationDate = &newExpDate
+
+			dormantGrants = append(dormantGrants, g)
+			dormantGrantsIDs = append(dormantGrantsIDs, g.ID)
+
+			dormantGrantsByOwner[g.Owner] = append(dormantGrantsByOwner[g.Owner], g)
+		}
+	}
+	s.logger.Info(fmt.Sprintf("found %d dormant grants for provider %q", len(dormantGrants), provider.URN), "grant_ids", dormantGrantsIDs)
+
+	if criteria.DryRun {
+		s.logger.Info("dry run mode, skipping updating grants expiration date")
+		return nil
+	}
+
+	if err := s.repo.BulkUpsert(ctx, dormantGrants); err != nil {
+		return fmt.Errorf("updating grants expiration date: %w", err)
+	}
+
+	go func() { // send notifications
```

Comment:
regardless sending notification is successful or error, we would still return `nil` error, so the function's return is not depending on it. Running in a goroutine just completely isolating and showing that it is independent and non-blocking

Reply:
Have we tested this go routine behaviour? 

I am afraid there will be some unwanted failure or behaviour since the preparation notification inside goroutine require some information from outside goroutine.

If there is a problem, maybe we could just put the `s.notifier.Notify` function only in goroutine and keep the message preparation outside.

Reply:
I see, I will just make it synchronous and test it out for now to avoid unwanted failure, since this part is also a little bit expensive to test